File: server/access.js
'use strict'

/**
 * Restricts the whiteboard to machines on the local network.
 *
 * This is a network boundary, not authentication: anyone who can join the same
 * Wi-Fi can reach the board. It keeps the internet out, nothing more.
 */

// RFC1918 private ranges, plus loopback and link-local.
const DEFAULT_IPV4_CIDRS = [
  '127.0.0.0/8',
  '10.0.0.0/8',
  '172.16.0.0/12',
  '192.168.0.0/16',
  '169.254.0.0/16',
]

/** Strips the IPv4-mapped IPv6 prefix and any zone id. */
function normalizeAddress(address) {
  if (typeof address !== 'string' || address.length === 0) return null
  let value = address.trim()
  const zone = value.indexOf('%')
  if (zone !== -1) value = value.slice(0, zone)
  if (value.startsWith('::ffff:')) value = value.slice('::ffff:'.length)
  return value.toLowerCase()
}

function parseIPv4(value) {
  const parts = value.split('.')
  if (parts.length !== 4) return null
  let result = 0
  for (const part of parts) {
    if (!/^\d{1,3}$/.test(part)) return null
    const byte = Number(part)
    if (byte > 255) return null
    result = (result << 8) | byte
  }
  return result >>> 0
}

function parseCidr(text) {
  const [network, bitsText] = String(text).trim().split('/')
  const base = parseIPv4(network)
  if (base === null) return null
  const bits = bitsText === undefined ? 32 : Number(bitsText)
  if (!Number.isInteger(bits) || bits < 0 || bits > 32) return null
  const mask = bits === 0 ? 0 : (0xffffffff << (32 - bits)) >>> 0
  return { base: (base & mask) >>> 0, mask }
}

function parseCidrList(text) {
  return String(text)
    .split(',')
    .map((entry) => entry.trim())
    .filter(Boolean)
    .map((entry) => {
      const cidr = parseCidr(entry)
      if (!cidr) throw new Error(`ALLOWED_CIDRS に不正な指定があります: "${entry}"`)
      return cidr
    })
}

/** Loopback and the private/link-local IPv6 ranges that mirror the IPv4 list. */
function isPrivateIPv6(value) {
  if (value === '::1' || value === '::') return true
  const prefix = value.slice(0, 2)
  // fc00::/7 (unique local) and fe80::/10 (link local).
  return prefix === 'fc' || prefix === 'fd' || prefix === 'fe'
}

/**
 * Builds the guard. `allowedCidrs` (comma separated, IPv4) replaces the default
 * private ranges; loopback stays allowed either way so the host machine and its
 * own health checks keep working.
 */
function createAccessGuard({ allowedCidrs } = {}) {
  const configured = allowedCidrs ? parseCidrList(allowedCidrs) : null
  const ranges = configured
    ? [parseCidr('127.0.0.0/8'), ...configured]
    : DEFAULT_IPV4_CIDRS.map(parseCidr)

  const isAllowed = (address) => {
    const value = normalizeAddress(address)
    if (!value) return false

    const ipv4 = parseIPv4(value)
    if (ipv4 === null) {
      // Custom ranges are IPv4 only, so IPv6 clients are held to loopback/ULA.
      return isPrivateIPv6(value)
    }
    return ranges.some(({ base, mask }) => ((ipv4 & mask) >>> 0) === base)
  }

  isAllowed.describe = () =>
    configured ? allowedCidrs : `プライベートアドレス全体 (${DEFAULT_IPV4_CIDRS.join(', ')})`

  return isAllowed
}

module.exports = { createAccessGuard, normalizeAddress, parseCidr, parseIPv4 }

File: server/app/access.py
"""Restricts the whiteboard to machines on the local network.

This is a network boundary, not authentication: anyone who can join the same
Wi-Fi can reach the board. It keeps the internet out, nothing more.
"""

from __future__ import annotations

from ipaddress import AddressValueError, ip_address, ip_network
from typing import Iterable

# RFC1918 private ranges, plus loopback and link-local, in both families.
DEFAULT_NETWORKS: tuple[str, ...] = (
    "127.0.0.0/8",
    "10.0.0.0/8",
    "172.16.0.0/12",
    "192.168.0.0/16",
    "169.254.0.0/16",
    "::1/128",
    "fc00::/7",
    "fe80::/10",
)

LOOPBACK_NETWORKS: tuple[str, ...] = ("127.0.0.0/8", "::1/128")


class InvalidRangeError(ValueError):
    """Raised at startup when ALLOWED_CIDRS cannot be parsed."""


def normalize_address(address: str | None) -> str | None:
    """Strips the IPv4-mapped IPv6 prefix and any zone id."""
    if not address:
        return None
    value = address.strip()
    if not value:
        return None
    zone = value.find("%")
    if zone != -1:
        value = value[:zone]
    if value.lower().startswith("::ffff:") and "." in value:
        value = value[len("::ffff:") :]
    return value


def _parse_networks(entries: Iterable[str]) -> list:
    networks = []
    for entry in entries:
        text = entry.strip()
        if not text:
            continue
        try:
            networks.append(ip_network(text, strict=False))
        except (ValueError, AddressValueError) as error:
            raise InvalidRangeError(
                f'ALLOWED_CIDRS に不正な指定があります: "{text}" ({error})'
            ) from error
    return networks


class AccessGuard:
    """Decides whether a connection may reach the board.

    `allowed_cidrs` (comma separated) replaces the default private ranges;
    loopback stays allowed either way so the host machine and its own health
    checks keep working.
    """

    def __init__(self, allowed_cidrs: str | None = None) -> None:
        self._configured = bool(allowed_cidrs and allowed_cidrs.strip())
        if self._configured:
            assert allowed_cidrs is not None
            self._description = allowed_cidrs.strip()
            self._networks = _parse_networks(LOOPBACK_NETWORKS) + _parse_networks(
                allowed_cidrs.split(",")
            )
        else:
            self._description = f"プライベートアドレス全体 ({', '.join(DEFAULT_NETWORKS)})"
            self._networks = _parse_networks(DEFAULT_NETWORKS)

    def allows(self, address: str | None) -> bool:
        value = normalize_address(address)
        if value is None:
            return False
        try:
            parsed = ip_address(value)
        except ValueError:
            return False
        return any(parsed in network for network in self._networks)

    def describe(self) -> str:
        return self._description
